### DISCUSSION TITLE
Need help creating abstract query wrapper in TypeScript

### DISCUSSION THREAD

**Discussion by piotrkulpinski** [trpc/trpc#5085] [Q&A] [answered]
Hi all!

I'm trying to create an abstracter wrapper to display some loading state indicators (plus other things), but I'm struggling with accessing the correct query types from passed queries.

Here's what I currently have:

```tsx
// app/page.tsx

export default function Page () {
  const { id: companyId } = useCompany()

  const queries = api.useQueries((t) =>[
    t.boards.getAll({ companyId }),
    t.statuses.getAll({ companyId }),
  ])

  return (
    <LoadingQueries queries={queries}>
      {(queriesData) => (
        <Subcomponent queriesData={queriesData} />
      )}
    </LoadingQueries>
  )
}
```

```tsx
// app/components/LoadingQueries.tsx

type LoadingQueriesProps<T> = PropsWithChildren<{
  queries: T // How to access the queries type here?
  children: (queriesData: UseBaseQueryResult<T>) => ReactNode // How to access the queries data array here?
}>

// The extend value is incorrect for sure
export const LoadingQueries = <T extends UseTRPCQueryResult<any, any>[]>({ children, queries }: LoadingQueriesProps<T>) => {
  const status = queries.reduce((acc, { status }) => {
    if (status === "loading") {
      return "loading"
    }

    if (status === "error") {
      return "error"
    }

    return acc
  }, "success" as QueryStatus)

  if (status === "error") {
    return null
  }

  return (
    <div>
      {status === "loading" ? <Loading /> : children(queries.map(({ data }) => data))}
    </div>
  )
}
```

Added some comments to show where I struggle with this
I would appreciate any help.

Thanks!

**Comment by Nick-Lucas** [comment]
If you're only using status and data why do you need the full types? Just type the component with exactly what you need and it gets way easier. `{ status: 'loading' | 'error', data: unknown }[]`

**Reply by piotrkulpinski** [reply]
I could do that, but then I loose all of the typings passed over to the `Subcomponent` which is not ideal.

**Reply by Nick-Lucas** [reply]
You only lose what you don’t implement, my example type was just a naive example. The trpc and RQ types can be a lot to wrangle so this is often the more practical way

**Comment by KATT** [comment, marked as answer]
See #1734

**Reply by piotrkulpinski** [reply]
This looks like exactly what I need thanks! It's a bummer it wasn't implemented. Seems like a great addition to the package.
Thanks
